<a href="https://colab.research.google.com/github/curttay1-dev/My-python-pandas-projects/blob/main/Pandas_Day_1_June_2026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# install library if needed
!pip install pandas

In [ ]:
# import pandas library
import pandas as pd
# alias pandas library as pd

In [ ]:
pd.__version__

'2.2.2'

# Creating a dataframe

In [ ]:
# list object
mylist = ["apple", "banana", "cherry"]

In [ ]:
# dataframe from a list
pd.DataFrame(mylist)

,0
0,apple
1,banana
2,cherry


In [ ]:
mydict = {"name":"Yusuf",
          "age":18,
          "location":"UK"}

In [ ]:
pd.DataFrame(mydict, index=[0])

,name,age,location
0,Yusuf,18,UK


# Import dataset

In [ ]:
df = pd.read_excel("/content/student.xlsx")
# import and assign to df
# df is short for dataframe

In [ ]:
print(type(df))

<class 'pandas.core.frame.DataFrame'>


In [ ]:
df = pd.read_csv("/content/student.csv")

# EDA (Exploratory Data Analaysis)

In [ ]:
df

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
# top 5 rows
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# top 10 rows
df.head(10)

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
# last 5 rows
df.tail()

,id,name,class,mark,gender
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
df.tail(3)

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
# random sample data
df.sample()

,id,name,class,mark,gender
12,13,Kty,Seven,88,female


In [ ]:
# random sample data
df.sample(3)

,id,name,class,mark,gender
3,4,Krish Star,Four,60,female
20,21,Babby John,Four,69,female
8,9,Tes Qry,Six,78,NaN


In [ ]:
# number of rows and columns
df.shape
#its an attribute. no ()

(35, 5)

In [ ]:
# dataset information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      35 non-null     int64 
 1   name    34 non-null     object
 2   class   34 non-null     object
 3   mark    35 non-null     int64 
 4   gender  33 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.5+ KB


- compare non-null count vs number of entries.
* name has 34 non-null
* class has 34 non-null
* gender has 33 non-null
**These columns have missing values.**

In [ ]:
# name of the columns
df.columns

Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [ ]:
# statistical summary
df.describe()

,id,mark
count,35.000000,35.000000
mean,18.000000,74.657143
std,10.246951,16.401117
min,1.000000,18.000000
25%,9.500000,62.500000
50%,18.000000,79.000000
75%,26.500000,88.000000
max,35.000000,96.000000


- compare mean vs std. std must be much less than mean. if its close or higher, there are outliers.
- 50% median. 50 % of the marks of the students are equal to higher than 79.
- 25%. 25% of all student have mark less than 62.5

In [ ]:
# number of rows
len(df)

35

In [ ]:
# rows * columns
df.size
# 35 * 5 = 175

175

In [ ]:
# datatypes of the columns
df.dtypes

,0
id,int64
name,object
class,object
mark,int64
gender,object


- object - string data type
- int64 - integer
- float64 - float

In [ ]:
# Counts missing values per column.
df.isnull().sum()

,0
id,0
name,1
class,1
mark,0
gender,2


In [ ]:
# Counts duplicate rows.
df.duplicated().sum()

np.int64(0)

# Slicing Dataframe

In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# use [] after df to select column or columns
df["name"]

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df["class"]

,class
0,Four
1,Three
2,Three
3,Four
4,Four
5,Four
6,Fifth
7,Five
8,Six
9,Four


In [ ]:
# use df.column_name
df.id

,id
0,1
1,2
2,3
3,4
4,5
5,6
6,7
7,8
8,9
9,10


In [ ]:
df.class

SyntaxError: invalid syntax (1584895042.py, line 1)

In [ ]:
df["class"].head()

,class
0,Four
1,Three
2,Three
3,Four
4,Four


In [ ]:
# average mark
df["mark"].mean()

np.float64(74.65714285714286)

In [ ]:
# median mark
df["mark"].median()

79.0

- you need to place name of the column in a list
- ["name", "class"] list of the columns

In [ ]:
# you need to double [] if you want to return more than 1 column
df[["name", "class"]].head()

,name,class
0,John Deo,Four
1,Max Ruin,Three
2,Arnold,Three
3,Krish Star,Four
4,John Mike,Four


In [ ]:
df[["name", "class", "mark"]].sample()

,name,class,mark
16,Tumyu,Six,54


# loc and iloc operators

In [ ]:
df.loc[1]

,1
id,2
name,Max Ruin
class,Three
mark,85
gender,male


In [ ]:
df.iloc[1]

,1
id,2
name,Max Ruin
class,Three
mark,85
gender,male


- df.loc[rows, column name]
- df.iloc[rows, colum index]

In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
# return name of the 1st student
df.loc[0, "name"]

'John Deo'

In [ ]:
# mark at student index 2
df.loc[2, "mark"]

np.int64(55)

- first row is index 0
- second row is index 1
- fist column is at index 0

In [ ]:
# return name of the 1st student
# df.loc[0, "name"]
# return name of the 1st student. name column is at index 1
df.iloc[0, 1]

'John Deo'

In [ ]:
# mark at student index 2
#df.loc[2, "mark"]
df.iloc[2, 3]

np.int64(55)

In [ ]:
# name, class of first 5 using loc
df.loc[0:4, ["name", "class"] ]

,name,class
0,John Deo,Four
1,Max Ruin,Three
2,Arnold,Three
3,Krish Star,Four
4,John Mike,Four


In [ ]:
# name, class of first 5 using iloc
df.iloc[0:5, [1,2]]
# iloc doesnt include the stopping point in the result

,name,class
0,John Deo,Four
1,Max Ruin,Three
2,Arnold,Three
3,Krish Star,Four
4,John Mike,Four


In [ ]:
df.loc[[0,5], ["name", "mark"]]

,name,mark
0,John Deo,75
5,Alex John,55


In [ ]:
df.iloc[[0,5], [1,3]]

,name,mark
0,John Deo,75
5,Alex John,55


In [ ]:
df.loc[10:15, ["name", "gender", "mark"]]

,name,gender,mark
10,Ronald,female,89
11,Recky,female,94
12,Kty,female,88
13,Bigy,female,88
14,Tade Row,male,88
15,Gimmy,male,88


In [ ]:
df.iloc[10:15, 1:3]

,name,class
10,Ronald,Six
11,Recky,Six
12,Kty,Seven
13,Bigy,Seven
14,Tade Row,NaN


# Filtering Dataframe

In [ ]:
df.loc[df["mark"]>90, ["name", "class", "mark"]]

,name,class,mark
11,Recky,Six,94
32,Kenn Rein,Six,96


In [ ]:
# df.iloc[df["mark"]>90, [1,2,3]]

In [ ]:
df.loc[df["class"]=="Six", :]
# : means return all range

,id,name,class,mark,gender
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
16,17,Tumyu,Six,54,male
29,30,Reppy Red,Six,79,female
32,33,Kenn Rein,Six,96,female
34,35,Rows Noump,Six,88,female


In [ ]:
df.loc[df["gender"]=="female", ["name", "gender"]]

,name,gender
0,John Deo,female
3,Krish Star,female
4,John Mike,female
9,Big John,female
10,Ronald,female
11,Recky,female
12,Kty,female
13,Bigy,female
19,Jackly,female
20,Babby John,female


- filter by multiple condition

In [ ]:
df.loc[(df["class"]=="Five") & (df["gender"]=="male"), ["name", "class", "gender"]]

,name,class,gender
7,Asruid,Five,male
17,Honny,Five,male


In [ ]:
df.loc[(df["class"]=="Seven") & (df["mark"]>=70), ["name", "class", "mark"]]

,name,class,mark
12,Kty,Seven,88
13,Bigy,Seven,88
23,Tiddy Now,Seven,78
24,Giff Tow,Seven,88
25,Crelea,Seven,79
27,Rojj Base,Seven,86
31,Binn Rott,Seven,90


In [ ]:
# write condition insidei [] to filter dataframe using slicing method
df[df["mark"]>85]

,id,name,class,mark,gender
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
14,15,Tade Row,NaN,88,male
15,16,Gimmy,Four,88,male
24,25,Giff Tow,Seven,88,male
27,28,Rojj Base,Seven,86,female
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female


In [ ]:
df[df["class"]=="Four"]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
9,10,Big John,Four,55,female
15,16,Gimmy,Four,88,male
20,21,Babby John,Four,69,female
30,31,Marry Toeey,Four,88,male


In [ ]:
df[(df["class"]=="Four") & (df["gender"]=="female")]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
20,21,Babby John,Four,69,female


In [ ]:
df[(df["class"]=="Four") |(df["gender"]=="female")]
# filter by class four or gender female

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
15,16,Gimmy,Four,88,male


In [ ]:
df.filter(items=['name', 'mark'])

,name,mark
0,John Deo,75
1,Max Ruin,85
2,Arnold,55
3,Krish Star,60
4,John Mike,60
5,Alex John,55
6,My John Rob,78
7,Asruid,85
8,Tes Qry,78
9,Big John,55


In [ ]:
df.filter(like='me')

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df.query("mark > 60")

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
14,15,Tade Row,NaN,88,male


In [ ]:
df.query('gender == "female"')

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
19,20,Jackly,Nine,65,female
20,21,Babby John,Four,69,female


# Sort Values

In [ ]:
df.sort_values(by="mark")

,id,name,class,mark,gender
18,19,Tinny,Nine,18,male
16,17,Tumyu,Six,54,male
2,3,Arnold,Three,55,male
5,6,Alex John,Four,55,male
28,29,Tess Played,Seven,55,male
21,22,Reggid,Seven,55,female
9,10,Big John,Four,55,female
4,5,John Mike,Four,60,female
3,4,Krish Star,Four,60,female
19,20,Jackly,Nine,65,female


In [ ]:
df.sort_values(by="name")

,id,name,class,mark,gender
5,6,Alex John,Four,55,male
2,3,Arnold,Three,55,male
7,8,Asruid,Five,85,male
20,21,Babby John,Four,69,female
9,10,Big John,Four,55,female
13,14,Bigy,Seven,88,female
31,32,Binn Rott,Seven,90,female
25,26,Crelea,Seven,79,male
33,34,Gain Toe,Seven,69,male
24,25,Giff Tow,Seven,88,male


In [ ]:
df.sort_values(by="mark", ascending=False)

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female
30,31,Marry Toeey,Four,88,male
34,35,Rows Noump,Six,88,female
24,25,Giff Tow,Seven,88,male
14,15,Tade Row,NaN,88,male
15,16,Gimmy,Four,88,male
12,13,Kty,Seven,88,female


In [ ]:
df.sort_values(by="mark", ascending=False).head().loc[df["gender"]=="female", :]

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female


In [ ]:
df[(df["gender"]=="female") | (df["class"]=="Seven")].sort_values(by="mark")

,id,name,class,mark,gender
9,10,Big John,Four,55,female
21,22,Reggid,Seven,55,female
28,29,Tess Played,Seven,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
19,20,Jackly,Nine,65,female
20,21,Babby John,Four,69,female
33,34,Gain Toe,Seven,69,male
0,1,John Deo,Four,75,female
23,24,Tiddy Now,Seven,78,male


In [ ]:
df[(df["gender"]=="female") | (df["class"]=="Seven")].sort_values(by=["id", "mark"])

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
19,20,Jackly,Nine,65,female
20,21,Babby John,Four,69,female


In [ ]:
## How many male and female students are there?
df["gender"].value_counts()

,count
gender,
male,17
female,16


In [ ]:
## Class and number of student
df["class"].value_counts()

,count
class,
Seven,10
Four,8
Six,7
Three,3
Nine,2
Five,2
Fifth,1
Eight,1
